crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,525 @@
+use std::collections::HashMap;
+
+use once_cell::sync::Lazy;
+
+use regex::RegexSet;
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit, Fix, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_text_size::{TextRange, TextSize};
+use rustpython_parser::lexer::LexResult;
+use rustpython_parser::Tok;
+
+use crate::{
+    registry::Rule,
+    settings::{flags, Settings},
+};
+
+/// ## What it does
+/// Checks that a TODO comment is actually labelled with "TODO".
+///
+/// ## Why is this bad?
+/// Ambiguous tags reduce code visibility and can lead to dangling TODOs.
+/// If someone greps for a TODO to fix, but the comment is tagged with a "FIXME"
+/// tag instead of a "TODO", that comment may never be found!
+///
+/// Note that this rule will only flag "FIXME", "BUG", and "XXX" tags as incorrect.
+///
+/// ## Example
+/// ```python
+/// # FIXME(ruff): this should get fixed!
+/// ```
+///
+/// Use instead:
+/// ```python
+/// # TODO(ruff): this is now fixed!
+/// ```
+#[violation]
+pub struct InvalidTodoTag {
+    pub tag: String,
+}
+impl Violation for InvalidTodoTag {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let InvalidTodoTag { tag } = self;
+        format!("Invalid TODO tag: `{tag}` should be `TODO`")
+    }
+}
+
+/// ## What it does
+/// Checks that a TODO comment has an author assigned to it.
+///
+/// ## Why is this bad?
+/// Assigning an author to a task helps keep it on the radar and keeps code
+/// formatting consistent.
+///
+/// ## Example
+/// ```python
+/// # TODO: should assign an author here
+/// ```
+///
+/// Use instead
+/// ```python
+/// # TODO(ruff): now an author is assigned
+/// ```
+#[violation]
+pub struct MissingAuthorInTodo;
+impl Violation for MissingAuthorInTodo {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Missing author in TODO. Try: # TODO (<author_name>): ...")
+    }
+}
+
+/// ## What it does
+/// Checks that an issue link or ticket is associated with a TODO.
+///
+/// ## Why is this bad?
+/// Including an issue link near a TODO makes it easier for resolvers
+/// to get context around the issue and keeps code formatting consistent.
+///
+/// ## Example
+/// ```python
+/// # TODO: this link has no issue
+/// ```
+///
+/// Use one of these instead:
+/// ```python
+/// # TODO (ruff): this comment has an issue link
+/// # https://github.com/charliermarsh/ruff/issues/3870
+///
+/// # TODO (ruff): this comment has a 3-digit issue code
+/// # 003
+///
+/// # TODO (ruff): this comment has an issue code of (up to) 6 characters, then digits
+/// # SIXCHR-003
+/// ```
+#[violation]
+pub struct MissingLinkInTodo;
+impl Violation for MissingLinkInTodo {
```

Comment:
I'm curious about this rule. It feels much more opinionated than the other rules and perhaps even unconventional (I haven't worked on a project with this requirement before). What are your thoughts on its relevance? I know it's part of the original implementation, but I wonder if this will be a rule that is `ignore`d more often than is healthy.

Reply:
It seems a bit different from the others, to be sure. I'd be okay with removing it. Seems like fixing it could take a decent chunk of work (i.e. making a new Github issue or something). Pretty much only included it since it was mentioned in the original `flake8-todo`.

Please let me know what you think! I think there'll be some additional work here because of the byte offset issue that Micha mentioned above, so happy to just make the same changes here.